hkmc2/shared/src/main/scala/hkmc2/codegen/HandlerLowering.scala

```diff
@@ -348,11 +350,106 @@ class HandlerLowering(using TL, Raise, Elaborator.State):
       case None => genNormalBody(b, BlockMemberSymbol("", Nil))
       case Some(cls) => Define(cls, genNormalBody(b, BlockMemberSymbol(cls.sym.nme, Nil)))
   
+  private val thirdPassFresh = FreshId()
+  // moves definitions to the top level of the block
+  private def thirdPass(b: Block): Block =
+
+    val (blk, defns) = popDefns(b, Nil)
+    val syms = defns.collect {
+      case ClsLikeDefn(sym, k, parentPath, methods, privateFields, publicFields, preCtor, ctor) => sym
+      case FunDefn(sym, params, body) => sym
+    }
+    val newSyms: Map[Symbol, Symbol] = syms.map {
+      case sym: BlockMemberSymbol => 
+        sym -> BlockMemberSymbol(sym.nme + "$" + thirdPassFresh(), sym.trees)
+      case sym: ClassSymbol => 
+        val newSym = ClassSymbol(sym.tree, Tree.Ident(sym.id.name + "$" +  + thirdPassFresh()))
+        newSym.defn = sym.defn
+        sym -> newSym
+      case sym: ModuleSymbol =>
+        sym -> ModuleSymbol(sym.tree, Tree.Ident(sym.id.name + "$" +  + thirdPassFresh()))
+      case sym => sym -> sym
+    }.toMap
+
+    val clsNmeMap = newSyms.collect {
+      case (c1: ClassSymbol, c2: ClassSymbol) => c1.id.name -> c2.id.name
+    }.toMap
+
+    def symMap(l: Local) = newSyms.get(l) match
+      case None => l match
+        case b: BlockMemberSymbol => clsNmeMap.get(b.nme) match
```

Comment:
Could you document the point of this `clsNmeMap`?

In principle, after the name resolution stage (done by `Elaborator`), there should be zero string-based resolution happening anywhere in the compiler. That's the purpose of symbols. Why do we need it here?

Reply:
The problem is the combination of the fact that `BlockMemberSymbol` is string-based, is used for class constructors, and is completely separated from `ClassSymbol`.

Every class has its own `ClassSymbol`. But they also have a separate `BlockMemberSymbol` for their constructor, which is currently purely string-based. Therefore, if we wish to change the name of a class, we are forced to update the name of the associated `BlockMemberSymbol` which involves changing the string.

The reason I need to change the name in the first place is to make them unique. I need to move function and class definitions to the top level of each instrumented function so that they can be accessed by the continuation. For example,
```
fun f() =
  if something then
    fun g() =
      fun h() = 2
      h()
    g()
  else
    fun g() = 3
    g()
```
will be changed to something like
```
fun f() =
  fun g1() =
      fun h() = 2
      h()
  fun g2() = 3
  if something then
    g1()
  else
    g2()
```
If two symbols are different but have the same name, they will be still emitted with the same name in the JS builder, so I need to rename the functions after moving them to the top level. If this is not the intended behaviour of the JS builder, then I can change it back after this is fixed.

Reply:
> The problem is the combination of the fact that `BlockMemberSymbol` is string-based

I guess by this you mean that it's "nominal", i.e., its name is visible/accessible to surface-language users and must be code-gen'd as is (without allocating a fresh name).

> is used for class constructors, and is completely separated from `ClassSymbol`.

This is by design, yes.

The issue is that so far, assuming nominality for that `BlockMemberSymbol` was fine. But since we now need to synthesize fresh `BlockMemberSymbol`, these should get fresh names allocated for them. Note that this fresh name allocation should happen in `JSBuilder` through the `Scope` construct. Manually dealing with creating fresh names in your transformation pass is unnecessary and suboptimal.

Let me see what's the best way of changing the current definitions to accommodate this.